src/engine/plugins/tile.cpp

```diff
@@ -181,20 +287,318 @@ Status TilePlugin::HandleRequest(const api::TileParameters &parameters, std::str
+        }
+
+        return offset;
+    };
+
+    // And a third time, should probably template this....
+    const auto use_point_float_value = [&used_point_floats,
+                                        &point_float_offsets](const float &value) {
+        const auto found = point_float_offsets.find(value);
+        std::size_t offset;
+
+        if (found == point_float_offsets.end())
+        {
+            used_point_floats.push_back(value);
+            offset = used_point_floats.size() - 1;
+            point_float_offsets[value] = offset;
+        }
+        else
+        {
+            offset = found->second;
+        }
+
+        return offset;
+    };
+
+    // If we're zooming into 16 or higher, include turn data.  Why?  Because turns make the map
+    // really
+    // cramped, so we don't bother including the data for tiles that span a large area.
+    if (parameters.z >= detail::MIN_ZOOM_FOR_TURNS)
+    {
+        // Struct to hold info on all the EdgeBasedNodes that are visible in our tile
+        // When we create these, we insure that (source, target) and packed_geometry_id
+        // are all pointed in the same direction.
+        struct EdgeBasedNodeInfo
+        {
+            unsigned source_intersection; // node-based-node ID
+            unsigned target_intersection; // node-based-node ID
+            unsigned packed_geometry_id;
+        };
+        // Lookup table for edge-based-nodes
+        std::unordered_map<NodeID, EdgeBasedNodeInfo> edge_based_node_info;
+        std::unordered_map<NodeID, std::vector<EdgeID>> outgoing_edges;
+        std::unordered_map<NodeID, std::vector<EdgeID>> incoming_edges;
+
+        // Now, loop over all the road segments we saw, and build up a mini
+        // graph for just the network that's visible.
+        for (const auto &edge : edges)
+        {
+            // Note: edge.u is the node-based node ID of the source intersection
+            //       edge.v is the node-based node ID of the target intersection
+            //       both these values can be directly looked up for coordinates
+
+            // If forward travel is enabled on this road section, and we haven't seen this
+            // edge-based-node
+            // before
+            if (edge.forward_segment_id.enabled &&
+                edge_based_node_info.count(edge.forward_segment_id.id) == 0)
+            {
+                // Add this edge-based-nodeid as an outgoing from the source intersection
+                auto f = outgoing_edges.find(edge.u);
+                if (f != outgoing_edges.end())
+                {
+                    f->second.push_back(edge.forward_segment_id.id);
+                }
+                else
+                {
+                    outgoing_edges[edge.u] = {edge.forward_segment_id.id};
+                }
+
+                // Add this edge-based-nodeid as an incoming to the target intersection
+                f = incoming_edges.find(edge.v);
```

Comment:
This is dangerous, since it's easy to compare the first map's iterator to the second map's iterator (which is UB) here; make separate var.